sourcemap/mapper.go

```diff
@@ -18,91 +18,33 @@
 package sourcemap
 
 import (
-	"fmt"
 	"strings"
-	"time"
 
-	logs "github.com/elastic/apm-server/log"
-
-	"github.com/elastic/beats/libbeat/common"
-	"github.com/elastic/beats/libbeat/logp"
+	"github.com/go-sourcemap/sourcemap"
 )
 
 const sourcemapContentSnippetSize = 5
 
-type Mapper interface {
-	Apply(Id, int, int) (*Mapping, error)
-	NewSourcemapAdded(id Id)
-}
-
-type SmapMapper struct {
-	Accessor Accessor
-	logger   *logp.Logger
+// Mapper holds a sourcemapper instance to be able to apply sourcemapping
+type Mapper struct {
```

Comment:
> I was going to suggest the full name for Mapper, but instead:
> Why the wrapper at all? What value is providing? 

I was thinking about that. But I appreciate encapsulating the concrete third party library used for sourcemapping in the actual `sourcemap` package. I don't see the advantage of spreading this out to the `model` package.

As stated above, I agree with the `error` handling change though.

Reply:
Again, all I can try to do is pointing out objectives arguments, which I didn't do in first comment. So:

Every object (and every name) is an abstraction per se, because they bring an extra level of indirection. Abstractions hide something and show something else instead. They are more complex (they are lies!), but are useful when they replace the _how_ something is done with _what_ is being done (so they are more declarative - "generic"). This way they hide something irrelevant (_how_) with something relevant (_what_), and information can be layered so you just can just look at only one layer any given time. 

I'm sure you know the theory, but this comes to explain that:

`Mapper` is not replacing _how_ with _what_, nor is extending the underlying `consumer` in any way, nor is encapsulating anything useful: `model/strackframe` is already importing `sourcemap` and it has the right data already (`lineno`/`colno`). `Apply` exists to hide the ugly arithmetic of the context lines (which is good!), but there is no need to hide the `consumer` from the `model` because there is not any implementation detail (_how_) leaking out; `consumer` has its own API that tells you _what_ is doing. 

Information is not laid out any better because of `mapper`. As a matter of fact, it introduces an extra indirection (more complex), a generic name (`mapper`) and redundancy: the code would be shorter without it, and the `nil` error check at the beginning of `Apply` could be removed (now the code does 2 `nil` checks: one for the `mapper`, and one for the `consumer`).

This is a good read re. abstractions in OOP if you are interested: https://simpleprogrammer.com/respecting-abstraction/

Reply:
Oh wow, thanks for the writeup.

I was actually misreading your suggestion, to move the whole `Apply` method directly to the stacktraceframe. That's why I argued with keeping the third party library in the `sourcemap` package. But yes, I can totally make this small change of not returning a wrapper around the consumer but directly the instance.